be/src/exprs/split_part.cpp

```diff
@@ -78,55 +73,89 @@ StatusOr<ColumnPtr> StringFunctions::split_part(FunctionContext* context, const
+                        offset = pos - haystack.data;
+                        num++;
+                    } else {
+                        offset = haystack.size;
+                        num = (num == 0) ? 0 : num + 1;
+                        break;
+                    }
                 }
-            }
 
-            if (num == part_number) {
-                res.append(Slice(haystack.data + pre_offset + 1, offset - pre_offset - 1));
+                if (num == part_number) {
+                    res.append(Slice(haystack.data + pre_offset + 1, offset - pre_offset - 1));
+                } else {
+                    res.append_null();
+                }
             } else {
-                res.append_null();
+                // if delimiter is a string, use memmem to split
+                int32_t pre_offset = -static_cast<int32_t>(delimiter.size);
+                int32_t offset = -static_cast<int32_t>(delimiter.size);
+                int32_t num = 0;
+                while (num < part_number) {
+                    pre_offset = offset;
+                    size_t n = haystack.size - offset - delimiter.size;
+                    char* pos = reinterpret_cast<char*>(
+                            memmem(haystack.data + offset + delimiter.size, n, delimiter.data, delimiter.size));
+                    if (pos != nullptr) {
+                        offset = pos - haystack.data;
+                        num++;
+                    } else {
+                        offset = haystack.size;
+                        num = (num == 0) ? 0 : num + 1;
+                        break;
+                    }
+                }
+
+                if (num == part_number) {
+                    res.append(
+                            Slice(haystack.data + pre_offset + delimiter.size, offset - pre_offset - delimiter.size));
+                } else {
+                    res.append_null();
+                }
             }
         } else {
-            // if delimiter is a string, use memmem to split
-            int32_t pre_offset = -static_cast<int32_t>(delimiter.size);
-            int32_t offset = -static_cast<int32_t>(delimiter.size);
+            part_number = -part_number;
+            auto haystack_str = haystack.to_string();
+            int32_t offset = haystack.size;
+            int32_t pre_offset = offset;
             int32_t num = 0;
-            while (num < part_number) {
-                pre_offset = offset;
-                size_t n = haystack.size - offset - delimiter.size;
-                char* pos = reinterpret_cast<char*>(
-                        memmem(haystack.data + offset + delimiter.size, n, delimiter.data, delimiter.size));
-                if (pos != nullptr) {
-                    offset = pos - haystack.data;
-                    num++;
+            auto substr = haystack_str;
+            while (num <= part_number && offset >= 0) {
+                offset = (int)substr.rfind(delimiter, offset);
```

Comment:
Could this implementation cause performance issues? Can it be implemented using memmem for reverse matching?